src/data_source/storage/sql.rs

```diff
@@ -1560,6 +1520,58 @@ where
             }
         }
 
+        // Reconstruct the merkle commitment from the path
+        let init = match proof_path.first() {
+            Some(MerkleNode::Leaf { value, .. }) => *value,
+            Some(MerkleNode::Empty) => State::T::default(),
+            Some(_) => {
+                return Err(QueryError::Error {
+                    message: "First node in the proof should be leaf or empty".to_string(),
+                })
+            }
+            None => return Err(QueryError::Missing),
+        };
+        let commitment_from_path = traversal_path
+            .iter()
+            .zip(proof_path.iter().skip(1))
+            .try_fold(init, |val, (branch, node)| -> QueryResult<State::T> {
+                match node {
+                    MerkleNode::Branch { value: _, children } => {
+                        let data = children
+                            .iter()
+                            .map(|node| match node.as_ref() {
+                                MerkleNode::ForgettenSubtree { value } => Ok(*value),
+                                MerkleNode::Empty => Ok(State::T::default()),
+                                _ => Err(QueryError::Error {
+                                    message: "Invalid child node".to_string(),
+                                }),
+                            })
+                            .collect::<QueryResult<Vec<_>>>()?;
+
+                        if data[*branch as usize] != val {
+                            return Err(QueryError::Error {
+                                message: "Some earlier state was used to calculate root commitment"
+                                    .to_string(),
+                            });
+                        }
+
+                        State::Digest::digest(&data).map_err(|_| QueryError::Missing)
+                    }
+                    MerkleNode::Empty => Ok(init),
```

Comment:
Isn't it an error if we encounter an empty node in the middle of the traversal path? I thought it should be branch nodes only for the internal nodes.

Reply:
There was a case where one index was removed and the non membership proof had all empty nodes

Reply:
Non membership can have empty nodes? I did remove one index from the tree and the proof had empty nodes

Reply:
Gotcha, this seems fine then